src/vs/editor/contrib/stickyScroll/browser/stickyScrollModelProvider.ts

```diff
@@ -0,0 +1,411 @@
+	 * @param textModelVersionId text-model version ID
+	 * @param token cancellation token
+	 * @returns the sticky model
+	 */
+	update(textModel: ITextModel, textModelVersionId: number, token: CancellationToken): Promise<StickyModel | null>;
+}
+
+export class StickyModelProvider implements IStickyModelProvider {
+
+	private _modelProviders: IStickyModelCandidateProvider[] = [];
+	private _modelPromise: CancelablePromise<any | null> | null = null;
+	private _updateScheduler: Delayer<StickyModel | null> | null = null;
+	private readonly _updateDebounceInfo: IFeatureDebounceInformation;
+	private readonly _store: DisposableStore;
+
+	constructor(
+		private _editor: ICodeEditor,
+		@ILanguageFeaturesService _languageFeaturesService: ILanguageFeaturesService,
+		@ILanguageConfigurationService _languageConfigurationService: ILanguageConfigurationService,
+		@ILanguageFeatureDebounceService _languageFeatureDebounceService: ILanguageFeatureDebounceService,
+		defaultModel: string) {
+
+		switch (defaultModel) {
+			case ModelProvider.OUTLINE_MODEL:
+				this._modelProviders.push(new StickyModelFromCandidateOutlineProvider(_languageFeaturesService));
+				this._modelProviders.push(new StickyModelFromCandidateSyntaxFoldingProvider(this._editor, _languageFeaturesService));
+				this._modelProviders.push(new StickyModelFromCandidateIndentationFoldingProvider(this._editor, _languageConfigurationService));
+				break;
+			case ModelProvider.FOLDING_PROVIDER_MODEL:
+				this._modelProviders.push(new StickyModelFromCandidateSyntaxFoldingProvider(this._editor, _languageFeaturesService));
+				this._modelProviders.push(new StickyModelFromCandidateIndentationFoldingProvider(this._editor, _languageConfigurationService));
+				break;
+			case ModelProvider.INDENTATION_MODEL:
+				this._modelProviders.push(new StickyModelFromCandidateIndentationFoldingProvider(this._editor, _languageConfigurationService));
+				break;
+		}
+
+		this._updateDebounceInfo = _languageFeatureDebounceService.for(_languageFeaturesService.foldingRangeProvider, 'Sticky Scroll Folding', { min: 200 });
+		this._store = new DisposableStore();
+	}
+
+	private _cancelModelPromise(): void {
+		if (this._modelPromise) {
+			this._modelPromise.cancel();
+			this._modelPromise = null;
+		}
+	}
+
+	public async update(textModel: ITextModel, textModelVersionId: number, token: CancellationToken): Promise<StickyModel | null> {
+
+		this._store.clear();
+		this._updateScheduler = new Delayer<StickyModel | null>(this._updateDebounceInfo.get(textModel));
+		this._store.add({
+			dispose: () => {
+				this._cancelModelPromise();
+				this._updateScheduler?.cancel();
+				this._updateScheduler = null;
+			}
+		});
+		this._cancelModelPromise();
+
+		return await this._updateScheduler.trigger(async () => {
+			const _stopWatch = new StopWatch(true);
+
+			for (const modelProvider of this._modelProviders) {
+				this._modelPromise = modelProvider.providerModelPromise;
```

Comment:
shouldn't that call be after `modelProvider.computeStickyModel`? `modelProvider.computeStickyModel` creates that promise.

Maybe modelProvider.computeStickyModel can return a `CancelablePromise`. Maybe `CancelablePromise<StickyModel | undefined>`, with `undefined` meaning it got cancelled.

Reply:
The issue is that the compute sticky model should also be able to return a third state which is NOT_VALID, indicating that the next model provider should be used. It could return for example: CancelablePromise<StickyModel | Status> where Status is an enum containing two values: CANCELED and NOT_VALID, but I wonder if this would make for _clean_ code.

Reply:
I changed the code so as to return an object containing the model promise as well as a status. The status (which has three possible states including undefined, valid and invalid) is necessary to determine if the code execution should entirely stop of continue to the next model provider.